### PR TITLE
Add "Ignore unused variable" code action

### PR DESCRIPTION
**Pull request by IMS94**
## Purpose
$subject

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var4.bal [hunk 1]
```diff
@@ -0,0 +1,9 @@
+type Person record {|
+    string name;
+    int id;
+|};
+
+function foo(Person person) returns int {
+    Person {id, name} = person;
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var4.bal:7]: The text edit produces a syntax error.
erro

**Review by malinthar — CHANGES_REQUESTED** [review]
changes look good!. I have put a couple of comments. Please address them if possible.

File: language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var4.bal (hunk 1 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var4.bal:7]: Thanks for pointing that out. My bad. Fixed the logic.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        } else {
+            return Collections.emptyList();
+        }
+
+        TextEdit textEdit = null;
+        if (bindingPatternNode.kind() == SyntaxKind.CAPTURE_BINDING_PATTERN) {
+            Range editRange = CommonUtil.toRange(bindingPatternNode.lineRange());
+            textEdit = new TextEdit(editRange, "_");
+        } else if (bindingPatternNode.kind() == SyntaxKind.FIELD_BINDING_PATTERN) {
+            if (bindingPatternNode instanceof FieldBindingPatternVarnameNode) {
```
[the remark this answers is not in the document]
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java:90]: Issue is, there are 2 sub classes of `FieldBindingPatternNode` named `FieldBindingPatternVarnameNode` and `FieldBindingPatternFullNode`. We can't differentiate them using the syntax kind. That's why I had to do the instance of check.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                Range editRange = CommonUtil.toRange(fieldBindingPattern.variableName().lineRange());
+                textEdit = new TextEdit(editRange, "_");
+            }
+        }
+
+        if (textEdit == null) {
+            return Collections.emptyList();
+        }
+
+        return List.of(createQuickFixCodeAction("Ignore unused variable", List.of(textEdit), context.fileUri()));
```
[the remark this answers is not in the document]
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java]: Yes. +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java:91]: Lets keep this as is in that case

**Review by malinthar — APPROVED** [review]
LGTM!

File: language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal [hunk 4]
```diff
@@ -0,0 +1,7 @@
+function calc(float a, float b) {
+    int[] arr = [1, 2, 3];
+    int total = 0;
+    foreach int i in arr {
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal:4]: I'm getting a syntax error inside the `foreach`.  Is this expected?

Screenshot 2021-11-03 at 09 49 33

Comment (TharushiJay) [reply language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal:4]: This was due to the diagnostics in the Compiler Toolkit. Therefore, marking as resolved.

Comment (TharushiJay) [reply language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal:4]: What is the expected behaviour when there are 2 unused variables?

Screenshot 2021-11-03 at 10 18 50

Comment (IMS94) [reply language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal:4]: I have the same question as well. Need to check with the front end team.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]
We need to gracefully handle the following scenarios. 

```
int pqr;
pqr = 1234;
```

```
int pqr = 14;
pqr = 12;
```

Once we accept the code action, this leads to a semantic error, which is not acceptable.

IMHO, we should show the code action only for the variable definitions when there are no usages. Otherwise we can't automatically remove the assignments.

File: language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal (hunk 4 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/test/resources/codeaction/ignore-unused-var/source/ignore_unused_var2.bal:4]: Updated the codeaction to ignore list binding pattern or mapping binding pattern in cases as above.

**Comment by IMS94** [comment]
@nadeeshaan added logic to not show this code action for variables with references. Can you check?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            TypedBindingPatternNode typedBindingPatternNode = (TypedBindingPatternNode) node;
+            bindingPatternNode = typedBindingPatternNode.bindingPattern();
+        } else if (node.kind() == SyntaxKind.CAPTURE_BINDING_PATTERN) {
+            bindingPatternNode = (BindingPatternNode) node;
+        } else if (node.kind() == SyntaxKind.FIELD_BINDING_PATTERN) {
+            bindingPatternNode = (BindingPatternNode) node;
+        } else if (node.kind() == SyntaxKind.MAPPING_BINDING_PATTERN) {
+            bindingPatternNode = (BindingPatternNode) node;
+        } else if (node.kind() == SyntaxKind.LIST_BINDING_PATTERN) {
+            bindingPatternNode = (BindingPatternNode) node;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java:94]: Can we collapse these check to one?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java (hunk 5 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/IgnoreUnusedVariableCodeAction.java]: Done!

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by malinthar — APPROVED** [review]
LGTM

**Pull request merged**